java/core/src/main/java/b3/TraceContext.java

```diff
@@ -0,0 +1,357 @@
+    }
+  }
+
+  public static final int FLAG_DEBUG = 1 << 0;
+  /** When set, we can interpret {@link #FLAG_SAMPLED} as a set value. */
+  public static final int FLAG_SAMPLING_SET = 1 << 1;
+  public static final int FLAG_SAMPLED = 1 << 2;
+  /**
+   * When set, we can ignore the value of the {@link #parentId}
+   *
+   * <p>While many zipkin systems re-use a trace id as the root span id, we know that some don't.
+   * With this flag, we can tell for sure if the span is root as opposed to the convention trace id
+   * == span id == parent id.
+   */
+  public static final int FLAG_IS_ROOT = 1 << 3;
+
+  TraceContext(TraceId traceId, long parentId, long spanId, long flags) {
+    this.traceId = traceId;
+    this.parentId = (parentId == spanId) ? traceId.lo : parentId;
+    this.spanId = spanId;
+    this.flags = flags;
+  }
+
+  /** Deserializes this from a big-endian byte array */
+  public static TraceContext fromBytes(byte[] bytes) {
+    checkNotNull(bytes, "bytes");
+    if (bytes.length != 32 && bytes.length != 40) {
+      throw new IllegalArgumentException("bytes.length " + bytes.length + " != 32 or 40");
+    }
+
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    long spanId = buffer.getLong(0);
+    long parentId = buffer.getLong(8);
+    TraceId traceId;
+    long flags;
+    if (bytes.length == 32) {
+      long lo = buffer.getLong(16);
+      traceId = new TraceId(0, (parentId == lo) ? parentId : lo);
+      flags = buffer.getLong(24);
+    } else {
+      traceId = new TraceId(
+          buffer.getLong(16),
+          buffer.getLong(24)
+      );
+      flags = buffer.getLong(32);
+    }
+
+    return new TraceContext(traceId, parentId, spanId, flags);
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Unique 8-byte identifier for a trace, set on all spans within it.
+   */
+  public final TraceId traceId;
+
+  /**
+   * The parent's {@link #spanId} or {@link #spanId} if this the root span in a trace.
+   */
+  public final long parentId;
+
+  /** Returns null when this is a root span. */
+  public Long nullableParentId() {
+    return root() ? null : parentId;
+  }
+
+  /**
+   * Unique 8-byte identifier of this span within a trace.
+   *
+   * <p>A span is uniquely identified in storage by ({@linkplain #traceId}, {@code #id}).
+   */
+  public final long spanId;
+
+  /** Returns true if this is the root span. */
+  public final boolean root() {
+    return (flags & FLAG_IS_ROOT) == FLAG_IS_ROOT || parentId == traceId.lo && parentId == spanId;
+  }
+
+  /**
+   * True is a request to store this span even if it overrides sampling policy. Implies {@link
+   * #nullableSampled()}.
+   */
+  public final boolean debug() {
+    return (flags & FLAG_DEBUG) == FLAG_DEBUG;
+  }
+
+  /**
+   * Should we sample this request or not? True means sample, false means don't, null means we defer
+   * decision to someone further down in the stack.
+   */
+  public final Boolean nullableSampled() {
```

Comment:
note to others.. I'm using "nullable" prefix naming convention. It is ugly, but anyone who treats this as not nullable has no excuse.